File: csharp/SwimmingCompetitionCS/networking/SwimmingCompetitionServicesJsonProxy.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading;
using model;
using Newtonsoft.Json;
using services;

namespace networking
{
    public class SwimmingCompetitionServicesJsonProxy : ISwimmingCompetitionServices
    {
        private string host;
        private int port;

        private ISwimmingCompetitionObserver client;
        private NetworkStream networkStream;
        private IFormatter formatter;
        private TcpClient connection;
        
        private Queue<Response> qresponses;
        private EventWaitHandle _waitHandle;
        private volatile bool finished;

        public SwimmingCompetitionServicesJsonProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            qresponses = new Queue<Response>();
        }

        public virtual void login(Organizer user, ISwimmingCompetitionObserver client)
        {
            Request req= JsonProtocolUtils.createLoginRequest(user);
            sendRequest(req);
            Response response=readResponse();
            if (response.Type== ResponseType.OK){
                this.client=client;
                return;
            }
            if (response.Type== ResponseType.ERROR)
            {
                string err = response.ErrorMessage;
                throw new SwimmingCompetitionException(err);
            }
        }

        public virtual Organizer findAccount(string username, string password)
        {
            initializeConnection();
            Request request =JsonProtocolUtils.createFindingAccountRequest(username, password);
            sendRequest(request);
            Response response = readResponse();
            if(response.Type == ResponseType.FINDING_ACCOUNT){
                Organizer organizer = response.Account;
                //Console.WriteLine(organizer.ToString());
                return organizer;
            }
            if (response.Type== ResponseType.ERROR){
                string err=response.ErrorMessage;
                closeConnection();
                throw new SwimmingCompetitionException(err);
            }
            return null;
        }

        public virtual ChallengeDTO[] findAllChallenges()
        {
            Request request =JsonProtocolUtils.createGetAllChallengesRequest();
            sendRequest(request);
            Response response = readResponse();
            if(response.Type == ResponseType.GET_ALL_CHALLENGES){
                ChallengeDTO[] challenges = response.AllChallenges;
                //Console.WriteLine(challenges);
                return challenges;
            }
            if (response.Type== ResponseType.ERROR)
            {
                string err = response.ErrorMessage;
                throw new SwimmingCompetitionException(err);
            }
            return null;
        }

        public virtual IList<ParticipantDTO> findParticipantsByChallenge(Challenge challenge)
        {
            Request request =JsonProtocolUtils.createFindingParticipantByChallengeRequest(challenge);
            sendRequest(request);
            Response response = readResponse();
            if(response.Type == ResponseType.FINDING_ALL_PARTICIPANTS_BY_CHALLENGE){
                IList<ParticipantDTO> participants = response.ParicipantsByChallenge;
                //Console.WriteLine(participants);
                return participants;
            }
            if (response.Type== ResponseType.ERROR){
                string err=response.ErrorMessage;
                throw new SwimmingCompetitionException(err);
            }

            return new List<ParticipantDTO>();
        }

        public virtual Participant addParticipant(string name, int age)
        {
            Request request =JsonProtocolUtils.createAddParticipantRequest(name, age);
            sendRequest(request);
            Response response = readResponse();
            if(response.Type == ResponseType.ADD_PARTICIPANT){
                Participant participant = response.AddedParticipant;
                //Console.WriteLine(participant);
                return participant;
            }
            if (response.Type== ResponseType.ERROR){
                string err=response.ErrorMessage;
                throw new SwimmingCompetitionException(err);
            }
            return null;
        }

        public virtual void addParticipantToMoreChallenges(Participant participant, List<Challenge> challenges)
        {
            Request request =JsonProtocolUtils.createAddParticipantToMoreChallengesRequest(participant, challenges);
            sendRequest(request);
            Response response = readResponse();
            if(response.Type == ResponseType.OK){
                //Console.WriteLine("Added participant " + participant + " to " + challenges);
            }
            if (response.Type== ResponseType.ERROR){
                string err=response.ErrorMessage;
                throw new SwimmingCompetitionException(err);
            }
        }

        public virtual void logout(Organizer user, ISwimmingCompetitionObserver client)
        {
            Request req=JsonProtocolUtils.createLogoutRequest(user);
            sendRequest(req);
            Response response=readResponse();
            closeConnection();
            if (response.Type== ResponseType.ERROR){
                String err=response.ErrorMessage;//data().toString();
                throw new SwimmingCompetitionException(err);
            }
        }

        private void closeConnection()
        {
            finished = true;
            try
            {
                networkStream.Close();
                connection.Close();
                _waitHandle.Close();
                client = null;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
            }
        }

        private void startReader()
        {
            Thread tw = new Thread(run);
            tw.Start();
        }

        private void sendRequest(Request request)
        {
            try
            {
                formatter.Serialize(networkStream, request);
                networkStream.Flush();
            }
            catch (Exception e)
            {
                throw new SwimmingCompetitionException("Error sending object " + e);
            }
        }

        private Response readResponse() {
            Response response= null;
            try{
                _waitHandle.WaitOne();
                lock(qresponses){
                    response = qresponses.Dequeue();
                }
            }
            catch (Exception e) {
                Console.WriteLine(e.Message);
            }

            return response;
        }
        
        private void initializeConnection(){
            try {
                connection=new TcpClient(host,port);
                this.networkStream = connection.GetStream();
                formatter = new BinaryFormatter();
                finished=false;
                _waitHandle = new AutoResetEvent(false);
                startReader();
            } 
            catch (IOException e) {
                Console.WriteLine(e.Message);
            }
        }
        private void handleUpdate(Response response){
            if(response.Type== ResponseType.UPDATE_TABLES){
                Console.WriteLine("Updating participants");
                client.updateTables(response.AllChallenges);
            }
        }
        private bool isUpdate(Response response){
            return response.Type == ResponseType.UPDATE_TABLES;
        }
        public virtual void run()
        {
            while(!finished)
            {
                try
                {
                    object response = formatter.Deserialize(networkStream);
                    //Console.WriteLine("response received "+response);
                    if (isUpdate((Response)response))
                    {
                        handleUpdate((Response)response);
                    }
                    else
                    {
							
                        lock (qresponses)
                        {
                            qresponses.Enqueue((Response)response);
                        }
                        _waitHandle.Set();
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine("Reading error "+e);
                }
					
            }
        }
    }
}

File: java/SwimmingCompetition/SwimmingCompetitionNetworking/src/main/java/swimmingcompetition/networking/json/SwimmingCompetitionServicesJsonProxy.java
package swimmingcompetition.networking.json;

import com.google.gson.Gson;
import swimmingcompetition.model.Challenge;
import swimmingcompetition.model.Organizer;
import swimmingcompetition.model.Participant;
import swimmingcompetition.model.Registration;
import swimmingcompetition.model.dto.ChallengeDTO;
import swimmingcompetition.model.dto.ParticipantDTO;
import swimmingcompetition.services.ISwimmingCompetitionObserver;
import swimmingcompetition.services.ISwimmingCompetitionServices;
import swimmingcompetition.services.SwimmingCompetitionException;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class SwimmingCompetitionServicesJsonProxy implements ISwimmingCompetitionServices {
    private String host;
    private int port;

    private ISwimmingCompetitionObserver client;

    private BufferedReader input;
    private PrintWriter output;
    private Gson gsonFormatter;
    private Socket connection;

    private BlockingQueue<Response> qresponses;
    private volatile boolean finished;
    public SwimmingCompetitionServicesJsonProxy(String host, int port) {
        this.host = host;
        this.port = port;
        qresponses=new LinkedBlockingQueue<Response>();
    }

    @Override
    public void login(Organizer user, ISwimmingCompetitionObserver client) throws SwimmingCompetitionException {
        //initializeConnection();

        Request req= JsonProtocolUtils.createLoginRequest(user);
        sendRequest(req);
        Response response=readResponse();
        if (response.getType()== ResponseType.OK){
            this.client=client;
            return;
        }
        if (response.getType()== ResponseType.ERROR){
            String err=response.getErrorMessage();;
            //closeConnection();
            throw new SwimmingCompetitionException(err);
        }
    }

    @Override
    public Organizer findAccount(String username, String password) throws SwimmingCompetitionException{
        initializeConnection();
        Request request =JsonProtocolUtils.createFindingAccountRequest(username, password);
        sendRequest(request);
        Response response = readResponse();
        if(response.getType() == ResponseType.FINDING_ACCOUNT){
            Organizer organizer = response.getAccount();
            //System.out.println(organizer.toString());
            return organizer;
        }
        if (response.getType()== ResponseType.ERROR){
            String err=response.getErrorMessage();;
            closeConnection();
            throw new SwimmingCompetitionException(err);
        }
        return null;
    }

    @Override
    public ChallengeDTO[] findAllChallenges() throws SwimmingCompetitionException{
        //initializeConnection();
        Request request =JsonProtocolUtils.createGetAllChallengesRequest();
        sendRequest(request);
        Response response = readResponse();
        if(response.getType() == ResponseType.GET_ALL_CHALLENGES){
            ChallengeDTO[] challenges = response.getAllChallenges();
            //System.out.println(challenges);
            return challenges;
        }
        if (response.getType()== ResponseType.ERROR){
            String err=response.getErrorMessage();;
            //closeConnection();
            throw new SwimmingCompetitionException(err);
        }
        return null;
    }

    @Override
    public List<ParticipantDTO> findParticipantsByChallenge(Challenge challenge) throws SwimmingCompetitionException{
        //initializeConnection();
        Request request =JsonProtocolUtils.createFindingParticipantByChallengeRequest(challenge);
        sendRequest(request);
        Response response = readResponse();
        if(response.getType() == ResponseType.FINDING_ALL_PARTICIPANTS_BY_CHALLENGE){
            List<ParticipantDTO> participants = response.getParicipantsByChallenge();
            //System.out.println(participants);
            return participants;
        }
        if (response.getType()== ResponseType.ERROR){
            String err=response.getErrorMessage();;
            //closeConnection();
            throw new SwimmingCompetitionException(err);
        }
        return new ArrayList<>();
    }

    @Override
    public Participant addParticipant(String name, int age) throws SwimmingCompetitionException{
       // initializeConnection();
        Request request =JsonProtocolUtils.createAddParticipantRequest(name, age);
        sendRequest(request);
        Response response = readResponse();
        if(response.getType() == ResponseType.ADD_PARTICIPANT){
            Participant participant = response.getAddedParticipant();
            //System.out.println(participant);
            return participant;
        }
        if (response.getType()== ResponseType.ERROR){
            String err=response.getErrorMessage();;
            //closeConnection();
            throw new SwimmingCompetitionException(err);
        }
        return null;
    }

    @Override
    public void addParticipantToMoreChallenges(Participant participant, List<Challenge> challenges) throws SwimmingCompetitionException{
        //initializeConnection();
        Request request =JsonProtocolUtils.createAddParticipantToMoreChallengesRequest(participant, challenges);
        sendRequest(request);
        Response response = readResponse();
        if(response.getType() == ResponseType.OK){
            //System.out.println("Added participant " + participant + " to " + challenges);
        }
        if (response.getType()== ResponseType.ERROR){
            String err=response.getErrorMessage();;
            //closeConnection();
            throw new SwimmingCompetitionException(err);
        }
    }

    private void closeConnection() {
        finished=true;
        try {
            input.close();
            output.close();
            connection.close();
            client=null;
        } catch (IOException e) {
            e.printStackTrace();
        }

    }
    public void logout(Organizer user, ISwimmingCompetitionObserver client) throws SwimmingCompetitionException {

        Request req=JsonProtocolUtils.createLogoutRequest(user);
        sendRequest(req);
        Response response=readResponse();
        closeConnection();
        if (response.getType()== ResponseType.ERROR){
            String err=response.getErrorMessage();//data().toString();
            throw new SwimmingCompetitionException(err);
        }
    }

    private void sendRequest(Request request)throws SwimmingCompetitionException {
        String reqLine=gsonFormatter.toJson(request);
        try {
            output.println(reqLine);
            output.flush();
        } catch (Exception e) {
            throw new SwimmingCompetitionException("Error sending object "+e);
        }

    }

    private Response readResponse() throws SwimmingCompetitionException {
       Response response=null;
        try{

            response=qresponses.take();

        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return response;
    }
    private void initializeConnection() throws SwimmingCompetitionException {
        try {
            gsonFormatter= new Gson();
            connection=new Socket(host,port);
            output=new PrintWriter(connection.getOutputStream());
            output.flush();
            input=new BufferedReader(new InputStreamReader(connection.getInputStream()));
            finished=false;
            startReader();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
    private void startReader(){
        Thread tw=new Thread(new ReaderThread());
        tw.start();
    }


    private void handleUpdate(Response response){
        if(response.getType()== ResponseType.UPDATE_TABLES){
            System.out.println("Updating tables");
            ChallengeDTO[] challenges = response.getAllChallenges();
            client.updateTables(challenges);
        }
    }

    private boolean isUpdate(Response response){
        return response.getType() == ResponseType.UPDATE_TABLES;
    }
    private class ReaderThread implements Runnable{
        public void run() {
            while(!finished){
                try {
                    String responseLine=input.readLine();
                    //System.out.println("response received "+responseLine);
                    Response response=gsonFormatter.fromJson(responseLine, Response.class);
                    if (isUpdate(response)){
                        handleUpdate(response);
                    }else{

                        try {
                            qresponses.put(response);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                } catch (IOException e) {
                    System.out.println("Reading error "+e);
                }
            }
        }
    }
}
